### PR TITLE
feat: support composing multiple http clients

### LINKED ISSUE #15: Allow InstabugHttpClient to be composable with other http clients
Our app integrates with several other services which provide custom http client implementations. In order to work with all of them we need a sensible way to compose the various clients together. One common way to do this is to provide an optional "inner client" that we can provide to which all network requests are forwarded. This would allow, for example, an application to integrate with DataDog in addition to Instabug.

### PR DESCRIPTION
**Pull request by dustin-graham**
## Description of the change
See #15 for details. Simply put, package consumers need a way to combine multiple http client implementations.
## Type of change
- [x] New feature (non-breaking change that adds functionality)

## Related issues

Fixes #15 
Jira ID: MOB-14744

### Development
- [x] Lint rules pass locally
- [x] The code changed/added as part of this pull request has been covered with tests
### Code review 
- [x] This pull request has a descriptive title and information useful to a reviewer
- [x] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,7 @@
+## [2.4.0]
```
Comment (a7medev) [inline CHANGELOG.md]: Let's change that to `Unreleased` as the version should be set on release.
```suggestion
## Unreleased
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,3 +1,7 @@
...
+
+- Enables InstabugHttpClient to wrap internal client
```
Comment (a7medev) [inline CHANGELOG.md]: Just a minor change.

```suggestion
- Enables `InstabugHttpClient` to wrap an internal `http` client.
```

File: example/lib/main.dart [hunk 3]
```diff
@@ -4,11 +4,13 @@ import 'package:instabug_flutter/instabug_flutter.dart';
...
-  Instabug.start(
-      'ed6f659591566da19b67857e1b9d40ab', [InvocationEvent.floatingButton]);
+  Instabug.init(
+    token: 'ed6f659591566da19b67857e1b9d40ab',
+    invocationEvents: [InvocationEvent.floatingButton],
+  );
   final client = InstabugHttpClient();
   final response = await client.get(Uri.parse('https://google.com'));
-  print(response.body);
+  debugPrint(response.body);
```
Comment (a7medev) [inline example/lib/main.dart]: I'd say let's move all analysis changes to a separate PR and make it the base for this one.

File: lib/instabug_http_client.dart [hunk 4]
```diff
@@ -10,11 +10,10 @@ import 'package:instabug_http_client/instabug_http_logger.dart';
 import 'package:meta/meta.dart';
 
 class InstabugHttpClient extends InstabugHttpLogger implements http.Client {
-  InstabugHttpClient() : client = http.Client() {
+  InstabugHttpClient({http.Client? client}) : client = client ?? http.Client() {
     logger = this;
   }
```
Comment (a7medev) [inline lib/instabug_http_client.dart]: The fact that both the parameter and the attribute are called `client` is a bit confusing, so I'd say let's change the class attribute to `_client` to avoid confusion.

**Review by a7medev — CHANGES_REQUESTED** [review]
Thanks @dustin-graham for opening this PR! Please check the review comments whenever you have time. 🙏🏼

File: lib/instabug_http_client.dart (hunk 4 above)
Comment (dustin-graham) [reply lib/instabug_http_client.dart]: @a7medev , I think I disagree with this recommendation. From an API perspective this is a completely expected Dart pattern. For example, if we were to write this constructor in a very common Dart way:
```dart
InstabugHttpClient({this.client}) {
  _logger - this.
}

final http.Client? client;
```
Consumers of this package would then be able to use the constructor in either of the following two ways:
```dart
// zero arg
InstabugHttpClient();
```
OR
```dart
// with custom inner client
InstabugHttpClient(client: myClient);
```

The way I've written the constructor in this PR allows for exactly the same usage.

Furthermore, changing the class variable to `_client` would be a breaking API change and absolutely would require a major version upgrade in pubspec.yaml. Given the previous code's @visibleForTesting annotation I think it would break tests. I recommend against this suggested change.

If we would like to distinguish between the InstabugHttpClient and the inner client provided by the caller, we could change the field to `innerClient` or something similar. However, this would still require a major version bump according to SemVer (non-backward compatible breaking change).

Please re-review. I think the API usage provided will feel natural to users and this implementation will not require a major version bump.

Comment (a7medev) [reply lib/instabug_http_client.dart]: @dustin-graham I think that's reasonable, let's continue with it this way then.

File: example/lib/main.dart (hunk 3 above)
Comment (a7medev) [reply example/lib/main.dart]: To not delay this PR further, let's keep them here.

File: pubspec.yaml [hunk 5]
```diff
@@ -2,7 +2,7 @@ name: instabug_http_client
 description:
   This package is an add on to instabug_flutter. It intercepts any requests performed
   with http Package and sends them to the report that will be sent to the dashboard.
-version: 2.3.0
+version: 2.4.0
```
Comment (a7medev) [inline pubspec.yaml]: @dustin-graham I think we should revert this change, this will be done once we release the package.

**Review by a7medev — CHANGES_REQUESTED** [review]
Great work @dustin-graham! This is a great addition to our `http` interceptor that will give users more flexibility. 🚀

There's only a small comment left, once resolved we can get it merged. 🙏🏼

File: pubspec.yaml (hunk 5 above)
Comment (dustin-graham) [reply pubspec.yaml]: @a7medev Sorry for the slow respomse. Now that the base package is 2.4.0, perhaps this is now OK?

**Comment by dustin-graham** [comment]
@a7medev we've been relying upon builds from our internal fork of this repo for a long time. I'd love to get this and another change we rely on integrated into this official repo soon. Please let me know if we can do this. Thanks!

File: lib/instabug_http_client.dart [hunk 6]
```diff
@@ -10,11 +10,15 @@ import 'package:instabug_http_client/instabug_http_logger.dart';
...
+  ///
+  /// Provide a value for [client] in order to override the internal client used
+  /// by this class. This can be useful if you are working with other libraries
+  /// that require other custom client implementations
+  InstabugHttpClient({http.Client? client}) : client = client ?? http.Client() {
     logger = this;
   }
-  @visibleForTesting
-  http.Client client;
+  final http.Client client;
```
Comment (a7medev) [inline lib/instabug_http_client.dart]: Is there a specific reason we're exposing the internal `client`? If not, it'd be better to leave it visible for testing.

```suggestion

  @visibleForTesting
  final http.Client client;
```

**Review by a7medev — CHANGES_REQUESTED** [review]
Hello @dustin-graham! Thanks for the contribution. 🙏🏼

I have a single comment and we can merge this PR.

File: lib/instabug_http_client.dart (hunk 6 above)
Comment (dustin-graham) [reply lib/instabug_http_client.dart]: I couldn't remember a reason for this, so I restored these lines. Thanks!

**Review by a7medev — APPROVED** [review]
LGTM 👌🏼
Thanks once again for the contribution @dustin-graham!

The PR has been merged and will be included in the next release. 🚀
